### PR TITLE
Ability to migrate with fragments

### PR DESCRIPTION
**Pull request by andreievg**
Fixes https://github.com/msupply-foundation/open-msupply/issues/3231

# 👩🏻‍💻 What does this PR do?

Although https://github.com/msupply-foundation/open-msupply/pull/4361 seems to deal with the issue of migrating RC versions and allows for dev help with migrations, I struggled to understand fully how it should be applied/used.

There were a lot of references to migration steps, which was perviously considered as a way to deal with multiple parallel migrations. It also deals with the two use cases that we have:

* Being able to migrate during RC
* Run migrations that were added in parallel (during feature, by differed devs)

Removed RC restrictions on migrations and added 'migration fragments' which are explained in the doc and example in this PR. But essentially a way to record and run migration 'steps' on top of migration by version. 

There was a question that migrations should be decoupled from version. I would agree with that if we have a mechanism of testing migrations, especially data migrations, by migration fragments. Although I think migration by versions does give a structure and reduces the chance of some old migrations being updated (expecting it to be auto applied ?)

## 💌 Any notes for the reviewer?

Please see readme.

Should we allow fragment migrations of previous versions ? (like if we did patch in main, should it also apply id dev)

# 🧪 Testing

See example in readme

# 📃 Documentation

- [ ] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

**Comment by mark-prins** [comment]
Took me a long time to understand the intention. Can I check that I have understood correctly?
* The current migration process remains: the `migration` fn is run if the migration version is greater than the database version. After migration the version in the database is updated, so the migration won't run again.
* There is a new fn added to the trait, which is run once only if the migration version is greater or equal to the database version (this is tracked by a record in a new table)

In which case: why not use the new table for _all_ migrations? Too many records in the table?

Side note: the 'migration fragments' are also one-time; so naming the 'migration' as 'one time' is not clear. I'm thinking in terms of 'major' and 'minor' migration aka semver; where you run the 'minor' when the major version is matching, and the 'major' migrations are a version bump. (though yes, it really is an RC migration, since major/minor/patch version changes all result in migrations. perhaps 'minor' is confusing things 🤔 )

**Comment by Chris-Petty** [comment]
weird

File: server/repository/src/migrations/README.md [hunk 1]
```diff
@@ -17,22 +17,58 @@ During startup server will run [these steps sequentially](mod.rs):
...
+
+1 - We start with adding a [new migration and a new table](https://github.com/msupply-foundation/open-msupply/compare/c95609f818f171bf106e8124c7ee87815d5f996e...4ede6643460459cb9a1bff4bbeea924bdb6c2e54).
+
+2 - [A field is added to that table](https://github.com/msupply-foundation/open-msupply/compare/4ede6643460459cb9a1bff4bbeea924bdb6c2e54...c53a65e19f1894baae7c2568aa8800fb69941a8d) (make sure it has reasonable defaults if it the column is not optional !). You can see that when devs are working in parallel on separate migrations, it should be possible to add consolidate migrations with these method, quite easily (since each fragment should be applied independently).
+
+3 - Modifying a field would require quite a lot of SQLite code, but we [can drop table and re-create it](https://github.com/msupply-foundation/open-msupply/compare/c53a65e19f1894baae7c2568aa8800fb69941a8d...71fbfb75a003f78cb29b678c6d1ade8810b98748).
+
+*Gotcha*
+
+Above quite works for simple cases, but it introduces too many use cases to work through, you should be at least aware of the surface area:
```
Comment (lache-melvin) [inline server/repository/src/migrations/README.md:61]: What is this sentence trying to say? 🙈 

```suggestion
The above works for simple cases, but it often introduces too many use cases to work through. You should be at least aware of the surface area:
```
?

File: server/repository/src/migrations/README.md [hunk 2]
```diff
@@ -17,22 +17,58 @@ During startup server will run [these steps sequentially](mod.rs):
...
+
+1 - We start with adding a [new migration and a new table](https://github.com/msupply-foundation/open-msupply/compare/c95609f818f171bf106e8124c7ee87815d5f996e...4ede6643460459cb9a1bff4bbeea924bdb6c2e54).
+
+2 - [A field is added to that table](https://github.com/msupply-foundation/open-msupply/compare/4ede6643460459cb9a1bff4bbeea924bdb6c2e54...c53a65e19f1894baae7c2568aa8800fb69941a8d) (make sure it has reasonable defaults if it the column is not optional !). You can see that when devs are working in parallel on separate migrations, it should be possible to add consolidate migrations with these method, quite easily (since each fragment should be applied independently).
+
+3 - Modifying a field would require quite a lot of SQLite code, but we [can drop table and re-create it](https://github.com/msupply-foundation/open-msupply/compare/c53a65e19f1894baae7c2568aa8800fb69941a8d...71fbfb75a003f78cb29b678c6d1ade8810b98748).
```
Comment (lache-melvin) [inline server/repository/src/migrations/README.md:57]: This compare is confusing... there's two copies of the `delete_after_rc_add_new_table` file, and the `delete_after_rc_drop_new_table` fragment is called twice??

However, want to chat about this recommendation - I think renaming the existing migration file, then creating a new one with the same name but a different identifier, feels pretty error prone 👀 Could we not just create another fragment called `update_data_to_double`, which is run after `add_new_table`, and in it we drop the table and recreate with the updated columns?

File: server/repository/src/migrations/README.md [hunk 3]
```diff
@@ -17,22 +17,58 @@ During startup server will run [these steps sequentially](mod.rs):
...
+
+Above quite works for simple cases, but it introduces too many use cases to work through, you should be at least aware of the surface area:
+
+If there is a field name change or field deletion or type change in a new branch, to which you switched and migrated. But then you switch back to develop for example, now we are in undefined state since rust logic will reference old fields and would expect old data types etc.
+
+If we added a reference to new table in another table, in parallel while between 2-3 were being worked on then migration will fail.
+
+Looking at 3 we may think it could be simplified by changing the identifier and [adding one drop clause](https://github.com/msupply-foundation/open-msupply/compare/c53a65e19f1894baae7c2568aa8800fb69941a8d...70827e95d03d359e80d967ac8e7ef29fe9ee72b3). This could work for in you in the branch you are working in, but then if you switch back to base branch, the previous migration fragments will try to be executed and will cause migration error
+
+Please be very mindful and vigilant when working with migration fragments, especially when doing major schema changes. And be aware of logic error that won't be caught at compile time.
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply server/repository/src/migrations/README.md:69]: +1 for the 'examples in code' comment; that was my key takeaway after reading the readme - would be great to have a code example. I think that'd be a more concise way to understand things

**Comment by lache-melvin** [comment]
Hi just updating here for transparency - discussed that we'll back out the example/recommendation of deleting some migrations after the RC phase. It may mean a few more mutations of the schema, but will be simpler to reason about. Once a migration fragment has been created, it should not be edited.

**Review by roxy-dao — APPROVED** [review]
Thanks! Just a carry over for the docs (:

**Comment by andreievg** [comment]
I created carry over issue to deal with docs.

Also added a commit to make sure RC3 is migrate-able: https://github.com/msupply-foundation/open-msupply/pull/4519/commits/6615973558a48e48b0e4412a455becdd1057a64f, re-ran all of the tests

Which looks at this diff
`git diff tags/v2.2.0-rc1 v2.2.0 server/repository/src/migrations/*`

```diff
diff --git a/server/repository/src/migrations/v2_02_00/rnr_form.rs b/server/repository/src/migrations/v2_02_00/rnr_form.rs
index f1ae19ac8f..c9d331a34a 100644
--- a/server/repository/src/migrations/v2_02_00/rnr_form.rs
+++ b/server/repository/src/migrations/v2_02_00/rnr_form.rs
@@ -39,6 +39,24 @@ pub(crate) fn migrate(connection: &StorageConnection) -> anyhow::Result<()> {
         "#
     )?;
 
+    #[cfg(feature = "postgres")]
+    sql!(
+        connection,
+        r#"
+          CREATE TYPE rn_r_form_low_stock AS ENUM (
+            'OK',
+            'BELOW_HALF',
+            'BELOW_QUARTER'
+          );
+        "#
+    )?;
+
+    const RNR_LOW_STOCK_ENUM_TYPE: &str = if cfg!(feature = "postgres") {
+        "rn_r_form_low_stock"
+    } else {
+        "TEXT"
+    };
+
     sql!(
         connection,
         r#"
@@ -46,6 +64,7 @@ pub(crate) fn migrate(connection: &StorageConnection) -> anyhow::Result<()> {
                 id TEXT NOT NULL PRIMARY KEY,
                 rnr_form_id TEXT NOT NULL REFERENCES rnr_form(id),
                 item_id TEXT NOT NULL REFERENCES item(id),
+                requisition_line_id TEXT REFERENCES requisition_line(id),
                 average_monthly_consumption {DOUBLE} NOT NULL,
                 previous_monthly_consumption_values TEXT NOT NULL,
 
@@ -63,6 +82,7 @@ pub(crate) fn migrate(connection: &StorageConnection) -> anyhow::Result<()> {
                 expiry_date {DATE},
                 calculated_requested_quantity {DOUBLE} NOT NULL,
                 entered_requested_quantity {DOUBLE},
+                low_stock {RNR_LOW_STOCK_ENUM_TYPE} NOT NULL,
 
                 comment TEXT,
                 confirmed BOOLEAN NOT NULL DEFAULT FALSE

```

**Pull request merged**